packages/salesforce/src/Utils.js

```diff
@@ -0,0 +1,144 @@
+    .catch(e => {
+      console.error(`Failed to connect to salesforce as ${username}`);
+      throw e;
+    });
+
+  console.info(`Connected to salesforce as ${username}.`);
+
+  return {
+    ...state,
+    connection,
+  };
+}
+
+function createAccessTokenConnection(state) {
+  const { instance_url, access_token } = state.configuration;
+
+  const connection = getConnection(state, {
+    instanceUrl: instance_url,
+    accessToken: access_token,
+  });
+
+  console.log(`Connected with ${connection._sessionType} session type`);
+
+  return {
+    ...state,
+    connection,
+  };
+}
+
+/**
+ * Creates a connection to Salesforce using Basic Auth or OAuth.
+ * @function createConnection
+ * @private
+ * @param {State} state - Runtime state.
+ * @returns {State}
+ */
+export function createConnection(state) {
+  if (state.connection) {
+    return state;
+  }
+
+  const { access_token } = state.configuration;
+
+  return access_token
+    ? createAccessTokenConnection(state)
+    : createBasicAuthConnection(state);
+}
+
+/**
+ * Removes state.connection from state.
+ * @example
+ * removeConnection(state)
+ * @function
+ * @private
+ * @param {State} state
+ * @returns {State}
+ */
+export function removeConnection(state) {
+  delete state.connection;
+  return state;
+}
+
+export async function pollJobResult(conn, job, pollInterval, pollTimeout) {
+  let attempt = 0;
+
+  const maxPollingAttempts = Math.floor(pollTimeout / pollInterval);
+
+  while (attempt < maxPollingAttempts) {
+    // Make an HTTP GET request to check the job status
+    const jobInfo = await conn
+      .request({
+        method: 'GET',
+        url: `/services/data/v${conn.version}/jobs/query/${job.id}`,
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      })
+      .catch(error => {
+        console.log('Failed to fetch job information', error);
+      });
+
+    if (jobInfo && jobInfo.state === 'JobComplete') {
+      const response = await conn.request({
+        method: 'GET',
+        url: `/services/data/v${conn.version}/jobs/query/${job.id}/results`,
+        headers: {
+          'Content-Type': 'application/json',
+        },
+      });
+
+      console.log('Job result retrieved', response.length);
+      return response;
+    } else {
+      // Handle maxPollingAttempts
+      if (attempt + 1 === maxPollingAttempts) {
+        console.error(
+          'Maximum polling attempt reached, Please increase pollInterval and pollTimeout'
+        );
+        throw new Error(`Polling time out. Job Id = ${job.id}`);
+      }
+      console.log(
+        `Attempt ${attempt + 1} - Job ${jobInfo.id} is still in ${
+          jobInfo.state
+        }:`
+      );
+    }
+
+    // Wait for the polling interval before the next attempt
+    await new Promise(resolve => setTimeout(resolve, pollInterval));
+    attempt++;
+  }
+}
+
+export const prepareNextState = (state, result, data) => {
```

Comment:
What is result? It looks new? How is it different to data?

Reply:
Well sometime we have result that comes with other metadata for example 
If you run `describeAll()` this will return 
`{"encoding": "UTF-8",  "maxBatchSize": 200, "sobjects": [{}, {}]}`. So the `sobjects` that can be `state.data` and `state.result` will be everything returned by salesforce `{"encoding": "UTF-8",  "maxBatchSize": 200, "sobjects": [{}, {}]}`

Reply:
But it's new though? This doesn't exist on main right now, that metadata is discarded.

The problem is that this prepareNextState signature doesn't make a lot of sense.

It would accept this, if you _really_ think it's valuable:
```
export const prepareNextState = (state, data, extras {}) => {
  return {
    ...composeNextState(state, data),
    ...extras,
  };
};
```

Reply:
I think we should make this simple and do this 
```
export const prepareNextState = (state, data) => {
  return {
    ...composeNextState(state, result),
    result,
  };
};
```
So the result will be in `state.data` and `state.result`.

Reply:
The code snippet isn't right (result is undefined). I assume you mean this:
```
export const prepareNextState = (state, result) => {
  return {
    ...composeNextState(state, result),
    result,
  };
};
```
In which case, you're just duplicating result and data? To what end?

Reply:
Now that i am thinking about it, we don't really need to add result. I think we have that pattern in our adaptor which adds `state.result` as well. 

I think the idea was even if that `state.data` gets updated you still have `state.result`

I guess this should be like this 👇🏽  then 
```
export const prepareNextState = (state, result) =>  composeNextState(state, result)
```

Reply:
I think `state.result` is an entirely new thing you've invented here :joy:  it's like `state.response` but I can't see it on main.

I vote to keep it simple and just do what's in your last comment

```
export const prepareNextState = (state, result) =>  composeNextState(state, result)
```

Reply:
I have updated `prepareNextState` function